rke_tests/common.py

```diff
@@ -0,0 +1,214 @@
+            cloud_provider.delete_node(node)
+
+
+def match_nodes(nodes, k8s_nodes):
+    """
+    Builds a list of tuples, where:
+    nodes_to_k8s_nodes[0][0] is the node object matched to
+    nodes_to_k8s_nodes[0][1] is the k8s info for the same node
+    TODO: what about etcd only nodes? are they returned in kubectl get nodes?
+    """
+
+    nodes_to_k8s_nodes = []
+    for k8s_node in k8s_nodes['items']:
+        hostname = k8s_node['metadata']['labels']['kubernetes.io/hostname']
+        for node in nodes:
+            if hostname == node.public_ip_address or \
+                    hostname == node.host_name:
+                nodes_to_k8s_nodes.append((node, k8s_node))
+    return nodes_to_k8s_nodes
+
+
+def assert_containers_exist_for_role(role, containers, expect_containers):
+    missing_containers = expect_containers[:]
+    for container in containers:
+        if container in expect_containers:
+            missing_containers.remove(container)
+    assert len(missing_containers) == 0, \
+        "Missing expected containers for role '{0}': {1}".format(
+            role, missing_containers)
+
+
+def validation_node_roles(nodes, k8s_nodes):
+    """
+    Validates each node's labels for match its roles
+    Validates each node's running containers match its role
+    Validates etcd etcdctl cluster-health command
+    Validates worker nodes nginx-proxy conf file for controlplane ips
+    """
+    # TODO: etcd only nodes are not listed in k8s_nodes, should handle this
+    # TODO: does hostname_override affect nginx-proxy controlplane ips?
+
+    role_matcher = {
+        'worker': 'node-role.kubernetes.io/worker',
+        'etcd': 'node-role.kubernetes.io/etcd',
+        'controlplane': 'node-role.kubernetes.io/master'}
+    controlplane = [
+        'scheduler', 'kube-controller', 'kubelet', 'kube-proxy', 'kube-api']
+    worker = ['kubelet', 'kube-proxy', 'nginx-proxy']
+    etcd = ['etcd']
+
+    controlplane_ips = []
+    for node in nodes:
+        if 'controlplane' in node.roles:
+            controlplane_ips.append(node.public_ip_address)
+
+    nodes_to_k8s_nodes = match_nodes(nodes, k8s_nodes)
+    for node, k8s_node in nodes_to_k8s_nodes:
+        for role in node.roles:
+            assert role_matcher[role] in k8s_node['metadata']['labels'].keys()
+            containers = node.docker_ps().keys()
+            if role == 'controlplane':
+                assert_containers_exist_for_role(
+                    role, containers, controlplane)
+            # nodes with work and controlplane roles do not have nginx-proxy
+            if role == 'worker' and 'controlplane' not in node.roles:
+                assert_containers_exist_for_role(role, containers, worker)
+                result = node.docker_exec(
+                    'nginx-proxy', 'cat /etc/nginx/nginx.conf')
+                for ip in controlplane_ips:
+                    assert 'server {0}:6443'.format(ip) in result, result
+            if role == 'etcd':
```

Comment:
Have checks for etcd nodes having non schedulable  kubelet containers .

Reply:
Now checks to verify taint is put on etcd only nodes